Native driver for Handy Tech braille displays

bramd opened the pull request:
### Link to issue number:
Closes  #7458.
fixes #5806.
closes #5418.

### Summary of the issue:
A rewrite of the driver for Handy Tech braille displays. This is a native implementation which doesn't depend on the Handy Tech COM server.

### Description of how this pull request fixes the issue:
Extensive discussion can be found in #7458. This is a first implementation of a native driver, which supports most Handy Tech displays. Notable changes for users include:

- Some very old displays have been dropped, but can be used via Handy Techs COM server and NVDA add-on
- Braille input is now available
- The Basic Braille displays will have working key bindings by default
- The Actilino display is supported
- Threadsafe, native driver which can benefit from things like braille display autodetection in the future

### Testing performed:

- I did extensive testing with a Modular Evolution 88 braille display
- @leonardder tested extensively with an Active Braille and did some testing with a Basic Braille 40 and a Basic Braille 80
- I did some testing with a Basic Braille 40 on USB and Bluetooth
- @zstanecic tested an earlier version with a Braille Wave

### Known issues with pull request:
- Support for some very old displays has been dropped
- This will need more work in the future to support advanced features like ATC, setting dot firmness etc

### Change log entry:

#### New features
* Support for the Actilino braille display from Handy Tech has been added
* Braille input for Handy Tech braille displays is now supported

#### Changes
* The Braillino, Bookworm and Modular (with old firmware) braille displays from Handy Tech are no longer supported out of the box. Install the Handy Tech Universal Driver and NVDA add-on to use these displays

bramd committed:
Basic native Handy Tech driver, work in progress. Outputs braille to Modular Evolution 88 and logs received packets and key presses

bramd committed:
Add model IDs and hardware identifiers for most displays

bramd committed:
Detect key releases and lgo them

bramd committed:
Basic Gesture for keys. Key tables per model need to be implemented. Routign is not yet supported.

bramd committed:
Cleanup imports

bramd committed:
Fix key release handling, convert _keysDown to a set, implemented cursor routing

Commit:
Add HID specific constants

Commit:
Allow sending HID packets

Commit:
Broken hid implementation, at least routing works

bramd committed:
Started work on model specific classes

bramd committed:
Merge remote-tracking branch 'babbage/handytech-native' into handytech-native

bramd committed:
Use cStringIO instead of StringIO

bramd committed:
Add classes for all models

Commit:
Another try to fix sending of HID packaets, not yet accounted for HID specific packets such as firmware version lookup

Commit:
Removed TripleActionKeysMixin from BasicBraille

Commit:
Moved the Actilino joystick keys into its own mixin, since ht plan to add a joystick to a new active braille at a later date. Also, fixed numcells, which should be 16

Commit:
No longer include the com server in the NVDA manifest

Commit:
Added support for spaces in Active Braille and other models

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd committed:
Map most keys to gestures

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd committed:
Add key table for BrailleWave and bind left/right for panning and esc/return for shift+tab/tab

bramd committed:
Correct class reference in super call for BrailleWave keys. Forgot to modify after copy-paste

bramd committed:
Correct class reference in super call in JoystickMixin.get_keys

bramd committed:
Fix pylint warnings

bramd committed:
Implemented braille input

Commit:
Added basic support for model specific gestures, currently not supported in gui

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd committed:
Improve regex for matching models, fix model specific gestures

bramd committed:
Delegate sending braille cells to the Model class, use AutoPropertyObject to rename get_keys to a keys property, some whitspace fixes, implemented old protocol for writing braille cells, construct name dynamically for models where this makes sense.

bramd committed:
No triple action keys for the Easy Braille

bramd committed:
Basic Braille specific gestures for scrolling left/right/up/down

bramd committed:
sendPacket and sendExtendedPacket are public methonds, since we use them from the Model classes

bramd committed:
eliminate some Pylint warnings

bramd committed:
Don't send model ID for simple packets of type braille

bramd committed:
Mixins that modify the keys property extend from AutoPropertyObject, so super calls usign the keys property work correctly. Enable ATC for capable models, so they can wake up from sleep on touch.

bramd committed:
Ignore ATC packets

Commit:
Fixed basic braille factory function, changed sendHIDPacket function and renamed some properties to use NVDA variable naming convensions

Commit:
DeviceID has now uppercase for ID, similar to the baum driver

Commit:
* Fix input gesture regex to also match properly against multiple keys
* Use the new protocol for easy braille again, since it seems to be supported quite well, even by fw 1.01
* Implemented support for left and right keys on easy braille, and allow to use them for braille input

Commit:
Ask for protocol properties, as the com server does this as well

Commit:
Fixed error raised with atc initialization without a model id

Commit:
Fix regex

Commit:
Implemented a postInit function for models

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd committed:
Add Bluetooth names

bramd committed:
Better postInit documentation, add pass in the base implementation

bramd committed:
Some code cleanup and parsing of protocol properties

 * Remove some OldProtocolMixin references
 * Untested parsing of the protocol properties, that seems to be given
   only by newer models
 * Raise a RuntimeError for unsupported displays

bramd committed:
Initialize HID serial converter (e.g. older devices using HID) properly (UNTESTED)

bramd committed:
Sets are not addable, use the intersection (|) operator instead

bramd committed:
Get rid of old protocol, add Modular Connect and Connect Braille 40,
account for status cells on the modular serie

bramd committed:
Drop support for the Bookworm

bramd committed:
Drop support for the Braillino for now

bramd committed:
Update documentation

 * Document supported displays in the user guide, some key definitions
   might still be missing
 * Remove references to the Handy Tech COM server from the readme

Commit:
Support HID packets with a size greater than writeSize

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd committed:
Merge remote-tracking branch 'babbage/handytech-native' into handytech-native

Commit:
Cache pending cells until received an ACK

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd committed:
Merge remote-tracking branch 'babbage/handytech-native' into handytech-native

bramd committed:
Update key assignments

bramd committed:
Revert model specific gestures for braille displays, already submitted as separate PR

bramd committed:
Update miscDeps to point to a commit without the Handy Tech COM server, remove reference to COM server from readme

bramd committed:
Revert braille.py correctly this time

bramd committed:
Merge branch 'handytech-native' of https://github.com/bramd/nvda into handytech-native

bramd committed:
Add translator comments

Commit:
Change variable names to  follow convensions

Commit:
Fixed some model classes with cells instead of numCells as their numCells variable

Commit:
Also for consistency, use Id in variable names instead of ID

derekriemer commented:
@bramd can you get us a link to this addon? It's not on the official repository, and we should add it when this merges.

zstanecic commented:
Hi,
the link is

ftp://ftp.handytech.de/public/nvda_braille_driver.nvda-addonftp://ftp.handytech.de/public/nvda_braille_driver.nvda-addonIt’s the handytech modification, so i don’t know, will we have the copyright issues @bramd

leonardder commented:
Try build for this pr is at https://ci.appveyor.com/api/buildjobs/4wd324ecwc1nrm9k/artifacts/output%2Fnvda_snapshot_try-handytech-native-14439%2C912cda4f.exe

Note that model specific modifiers are not yet supported in this try build, as this is covered by a separate pr, which is likely to be merged first. I will produce a new try build when that happens.

zstanecic commented:
Hi @leonardder, 
After downloading and testing the try build,
i can say that we have a lot of positive inpact on performance in microsoft word.
Braille performance in MS Word is faster.
Braille input works'
The keymap is working'
The driver throws no errors

leonardder commented:
Ugh, I forgot to point out that this try build will break your config with regard to the handy tech driver, it won't allow you to load the old driver any more, unless you manually change your nvda.ini and remove the handytech section from there. This is actually a problem that will be fixed by #1271 as well.

zstanecic commented:
Note:
I have tested this using the portable try build

Wysłane z aplikacji Poczta dla Windows 10

Od: Leonard de Ruijter
Wysłano: sobota, 16 września 2017 12:40
Do: nvaccess/nvda
DW: zstanecic; Mention
Temat: Re: [nvaccess/nvda] Native driver for Handy Tech braille displays(#7590)

Ugh, I forgot to point out that this try build will break your config with regard to the handy tech driver, it won't allow you to load the old driver any more, unless you manually change your nvda.ini and remove the handytech section from there. This is actually a problem that will be fixed by #1271 as well.
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

leonardder commented:
@FelixGruetzmacher: Your code review of this would be greatly appreciated.

Commit:
Support the hid converter properly this time

FelixGruetzmacher commented:
Did a detailed analysis and it mostly looks great.

## Issues I found and fixed

- Added status routing keys for Modular series.
- Fixed implementation for adding status cells for modular series. This tried to contatenate str to list.
- Fixed joystick. This didn't work because anything >= KEY_ROUTING was interpreted as routing, and the joystick comes after those.

I am attaching the version I am testing right now, which contains the fixes.

## Open issues to discuss

- Tried planting the driver into an installed copy of 2017.3, and reverted config to factory defaults. No trouble with HID devices, but with a serial device I received erroneous packets. Could it be that some underlying part of the implementation changed post 2017.3? Something in serial? Note: I double-checked that the COM server wasn't interfering.
- On a Win 10 machine I got a unicode translation error when it tried to get the available ports. Will give more details later---am away from that machine.
- What should we do about lost connections? Periodically re-detect in the driver? Any plans for that kind of infrastructure in core?

That's it for now. I'm very enthusiastic about this implementation, in case it didn't show!
[handyTech.zip](https://github.com/nvaccess/nvda/files/1341224/handyTech.zip)

leonardder commented:
@FelixGruetzmacher: could you file a pr against handytech-native brams repo, or do you want one of us to go over it and merge?

FelixGruetzmacher commented:
I'll clone and file a PR. I'm quite new to GIT, so I must apologize if I'm
not yet adjusted to the workflows.

Leonard de Ruijter <notifications@github.com> schrieb am Do., 28. Sep. 2017
um 17:20 Uhr:

> @FelixGruetzmacher <https://github.com/felixgruetzmacher>: could you file
> a pr against handytech-native brams repo, or do you want one of us to go
> over it and merge?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-332870526>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KMxL6q4_BEVSb_uEDPNWja6lOIqLks5sm7k4gaJpZM4PT5As>
> .
>

leonardder commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [28 Sep 2017, 19:23 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-332906098 "2017-09-28T17:23:56Z - Replied by Github Reply Comments"):
> I'll clone and file a PR. I'm quite new to GIT, so I must apologize if I'm
> not yet adjusted to the workflows.

No problem, this is not quite usual anyway. What you basically do is clone https://github.com/bramd/nvda, checkout handytech-native, commit your changes and than push it to your own fork. Than, from your own fork's branches page, create a pr against bramd:handytech-native.
Note that I pushed a commit this afternoon, so stuff might be out of sync now. It's no problem for me to do the job, so feel free to give a shout if it is a bit too scary.

bramd commented:
@FelixGruetzmacher Thanks a lot for your feedback. Before the last change of @leonardder, there were issues with the serial-HID converters, he tested a BrailleStar 40 with HID interface. However, I'm not aware of any issues with the serial communication. To get detailed logging of every packet being send/received add the following to your nvda.ini and set your log level to debug:

```
[debugLog]
	hwIo = True
```

Also, automatic braille display detection is discussed in issue #1271. That would include detecting the same display as before (e.g. after losing connection or switching communication method). That's also why I let the driver just fail when another model ID is detected, that will trigger the auto detection procedure in the future and just re-initialize the driver with the right parameters.

If you have questions about the git workflow, feel freee to mail/call me.

bramd commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -517,6 +517,7 @@ def __init__(self, port="auto"):
 		self._brailleInput = False
 		self._pendingCells = []
 		self._awaitingACK = False
+		self.hidSerialBuffer = ""
```
I would suggest starting this variable name with an _, it's private after all

Commit:
Review action: hidSerialBuffer is now a private variable

FelixGruetzmacher commented:
Here is some more about the error I get when using it in installed 2017.3 rather than the try build. I get the following in the log:

DEBUG - hwIo.Serial.read (09:11:56.572):
Read: ''

So it's reading an empty string. It never does this in the try build. Any recent changes to hwIo.Serial which might account for this?

To be on the safe side, I even renamed the com server so it wouldn't load no matter what, and also unregistered it so the manifest wouldn't pick it up anywhere else.

The driver, to my mind, does the right thing. It detects something's strange, empties the Braille cells and gracefully closes the port. It's just that I can't explain where that empty string comes from. This looks like something that should never happen.

FelixGruetzmacher committed:
Added status routing keys for modular series.
Fixed status cells for Modular series.
Fixed joystick detection by making the condition for a routing key more
restrictive.

bramd commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [Sep 29, 2017, 9:26 AM GMT+2](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333052143 "2017-09-29T07:26:19Z - Replied by Github Reply Comments"):
> So it's reading an empty string. It never does this in the try build. Any recent changes to hwIo.Serial which might account for this?

I can't think of any, but will have to diff the release and latest hwIo to be sure. For context, which display is this? And what kind of communication is happening before the reading of the empty string?

FelixGruetzmacher commented:
It's a Modular 80. The context is we've just gotten a positive acknowledge
for a display update, then out of the blue comes this empty string and the
driver is seeing another 0x16, probably because the input buffer wasn't
updated properly behind the scenes.
By the way, created a PR for my fixes. Hope I did it right. ;)

Bram Duvigneau <notifications@github.com> schrieb am Fr., 29. Sep. 2017 um
11:29 Uhr:

> *@FelixGruetzmacher* <https://github.com/FelixGruetzmacher> commented on Sep
> 29, 2017, 9:26 AM GMT+2
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333052143>:
>
> So it's reading an empty string. It never does this in the try build. Any
> recent changes to hwIo.Serial which might account for this?
>
> I can't think of any, but will have to diff the release and latest hwIo to
> be sure. For context, which display is this? And what kind of communication
> is happening before the reading of the empty string?
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333078745>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KAk9RbWAeSBJP12ayE6fdT0IL9tSks5snLhfgaJpZM4PT5As>
> .
>

derekriemer commented:
wow guys, way to go for collaboration on this.

FelixGruetzmacher commented:
Updated to Master, problem is gone. So no need to go phantom-hunting, it's
solved, whatever it was, and will propagate to stable anyway.

Derek Riemer <notifications@github.com> schrieb am Fr., 29. Sep. 2017 um
11:55 Uhr:

> wow guys, way to go for collaboration on this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333084848>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KKS_2DMPRzC0pe3get86655nWeGXks5snL6HgaJpZM4PT5As>
> .
>

Commit:
Fix unhandled exception when trying to destroy a Hid object without a file handle

Commit:
Fix a decode error for hwPortUtils logging when enumerating serial ports.

bramd committed:
Don't use an assert to read the extended packet terminator. The assert never gets called in release builds and this causes a driver crash with serial devices

bramd committed:
Merge remote-tracking branch 'upstream/master' into handytech-native

bramd commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [Sep 29, 2017, 1:15 PM GMT+2](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333099814 "2017-09-29T11:15:26Z - Replied by Github Reply Comments"):
> <div class="email-fragment">Updated to Master, problem is gone. So no need to go phantom-hunting, it's
> solved, whatever it was, and will propagate to stable anyway.

No, it's not. This was a good catch, thanks for testing on a release build. We used an assert to read the extended packet terminator "\x16". However, an assert never gets called on a release build. This was not a problem with native HID devices, since they deliver packets to the onReceive callback, but serial devices work like a stream. So, if the terminator byte isn't read, the "\x16" is send to onReceive as the start of a new packet and the whole thing goes down.

bramd committed:
Merge pull request #1 from FelixGruetzmacher/handytech-native

Fixes for Actilino and Modular

bramd committed:
Slightly changed if statement for routing keys

Commit:
* Rename StatusRoutingMixin to StatusCellMixin
* Move display method from Modular class to StatusRoutingMixin
* Assign joystick to arrow keys/enter key
* Add Felix to driver copyright

leonardder commented:
@MichaelDCurran: Especially now @FelixGruetzmacher from Handy Tech gave his review approval to this code, I think this is ready for review. May I ask whether it is possible to prioritise the review of this pr in a way that it can be merged into master before the release of 2017.4? One of the major reasons for this request is that this driver adds support for the new Handy Tech Actilino device. Supporting this device requires an update of the com server otherwise.

michaelDCurran commented:
I will review this now. However, the most important question I have is: A part from adding support for more models, does this cause any other user-visible changes? E.g. Is the new driver the same name and therefore will automatically be used with  previous configurations?  Do any features not work that used to?

michaelDCurran commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -3,130 +3,831 @@
+
+
+# Model dict for easy lookup
+def _allSubclasses(cls):
```
Please clearly explain exactly what this function is doing.

michaelDCurran commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -3,130 +3,831 @@
+
+	def __init__(self, display):
+		super(Model, self).__init__()
+		self._display = display
```
Setting _display to display and _model to model causes a circular reference. If __del__ is ever implemented on either of those classes, the garbage collector will not be able to clean them up.  I don't think this is happening right now. But it may be safer to make _display a weakreference. This should be okay as model instances only ever exist within the lifetime of a display.

michaelDCurran reviewed (changes requested):
This looks very good.
Two extra queries other than the other inline comments I have made:
1. log.warning is used a bit. Should that instead be debugWarning? Or  was this deliberate?
2. There are a few comments about things to do. Do these need to be done before this goes to Next?

leonardder commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [3 Oct 2017, 02:39 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333705237 "2017-10-03T00:39:41Z - Replied by Github Reply Comments"):
> I will review this now. However, the most important question I have is: A part from adding support for more models, does this cause any other user-visible changes? E.g. Is the new driver the same name and therefore will automatically be used with  previous configurations?  Do any features not work that used to?

Yes, the driver has the same name and thus will be used automatically. It lacks the following features:

1. No access to the handy tech control panel, as that was part of the com server. This panel allowed you to set the number of status cells, enable/disable active tactile control (which wasn't used anyway) and check some device info such as the serial number and dealer information
2. There is no support for the bookworm device (very old model) and the Braillino (also a model which is quite old, and due to the limited amount of 20 cells, probably hardly to never used in combination with NVDA nowadays). For the latter, it might be good to describe a proper workaround, at least in the whats new. That work around would be using a custom driver by handy tech. @FelixGruetzmacher knows more about that. The reason why these devices aren't supported is that they use a different protocol.

derekriemer commented:
Will this support dot firmness and active displays? Just curious.

leonardder commented:
[**@derekriemer**](https://github.com/derekriemer) commented on [3 Oct 2017, 06:06 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333733099 "2017-10-03T04:06:01Z - Replied by Github Reply Comments"):
> Will this support dot firmness and active displays? Just curious.

It supports the active line of displays, but currently it ignores all packets related to ATC, except for waking the display up by touching it, which is an internal feature of the ATC displays when ATC is on. It does not yet support dot firmness, as at least in my opinion, this requires #7452 to be implemented to some extend.

FelixGruetzmacher commented:
I'll soon get around to working on ATC support for this.

Leonard de Ruijter <notifications@github.com> schrieb am Di., 3. Okt. 2017
um 06:09 Uhr:

> *@derekriemer* <https://github.com/derekriemer> commented on 3 Oct 2017,
> 06:06 CEST
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333733099>:
>
> Will this support dot firmness and active displays? Just curious.
>
> It supports the active line of displays, but currently it ignores all
> packets related to ATC, except for waking the display up by touching it,
> which is an internal feature of the ATC displays when ATC is on. It does
> not yet support dot firmness, as at least in my opinion, this requires
> #7452 <https://github.com/nvaccess/nvda/issues/7452> to be implemented to
> some extend.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333733433>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KAjl8li05IjqwKRMQQAnbtL5ys80ks5sobNjgaJpZM4PT5As>
> .
>

leonardder commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [3 okt. 2017 10:41 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333777852 "2017-10-03T08:41:58Z - Replied by Github Reply Comments"):
> I'll soon get around to working on ATC support for this.

I assume you mean this to happen in a separate pr?

bramd committed:
Merge branch 'handytech-native' of github.com:bramd/nvda into handytech-native

bramd commented on source/hwIo.py:
```diff
@@ -97,7 +97,11 @@ def close(self):
+			self.close()
+		except:
+			if _isDebug():
+				log.debugWarning("Couldn't delete object gracefully", exc_info=True)
```
@leonardder You made this change to catch an exception that happened when closing a HID device without a file handle (according to your commit message). I don't have devices where this happens. Wouldn't it be better to check if we have a read/write handle in the close method and handle this specific situation there? That would also be more in line how hwIo.Serial does it.

leonardder replied:
Sure, that's also an option and I think I prefer this approach. Thanks

bramd commented:
[**@leonardder**](https://github.com/leonardder) commented on [Oct 3, 2017, 6:01 AM GMT+2](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333732570 "2017-10-03T04:01:37Z - Replied by Github Reply Comments"):
> [**@michaelDCurran**](https://github.com/michaelDCurran) commented on [3 Oct 2017, 02:39 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-333705237):
> Yes, the driver has the same name and thus will be used automatically. It lacks the following features:
> 
> 1.  No access to the handy tech control panel, as that was part of the com server. This panel allowed you to set the number of status cells, enable/disable active tactile control (which wasn't used anyway) and check some device info such as the serial number and dealer information

I think most info and functionality of this control panel will return when we have display specific settings in the GUI.

> 2.  There is no support for the bookworm device (very old model) and the Braillino (also a model which is quite old, and due to the limited amount of 20 cells, probably hardly to never used in combination with NVDA nowadays). For the latter, it might be good to describe a proper workaround, at least in the whats new. That work around would be using a custom driver by handy tech. [@FelixGruetzmacher](https://github.com/felixgruetzmacher) knows more about that. The reason why these devices aren't supported is that they use a different protocol.

By the way, a link to Handy Techs NVDA page is already provided in the user guide.

An additional issue I see is that the driver has a port parameter and the previous driver did not. So, people installing this and downgrading to an NVDA release containing the COM based driver will get an error and need to change nvda.ini manually. I have no idea how we can prevent this.

FelixGruetzmacher commented:
How about renaming the new driver to handyTechNative or something along
these lines? Then downgrading would never become a problem, and there would
not be any gesture conflicts.

Bram Duvigneau <notifications@github.com> schrieb am Di., 3. Okt. 2017 um
23:27 Uhr:

> *@leonardder* <https://github.com/leonardder> commented on Oct 3, 2017,
> 6:01 AM GMT+2
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333732570>:
>
> *@michaelDCurran* <https://github.com/michaelDCurran> commented on 3 Oct
> 2017, 02:39 CEST
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333705237>:
>
>
> Yes, the driver has the same name and thus will be used automatically. It
> lacks the following features:
>
>
>    1. No access to the handy tech control panel, as that was part of the
>    com server. This panel allowed you to set the number of status cells,
>    enable/disable active tactile control (which wasn't used anyway) and check
>    some device info such as the serial number and dealer information
>
> I think most info and functionality of this control panel will return when
> we have display specific settings in the GUI.
>
>
>    1. There is no support for the bookworm device (very old model) and
>    the Braillino (also a model which is quite old, and due to the limited
>    amount of 20 cells, probably hardly to never used in combination with NVDA
>    nowadays). For the latter, it might be good to describe a proper
>    workaround, at least in the whats new. That work around would be using a
>    custom driver by handy tech. @FelixGruetzmacher
>    <https://github.com/felixgruetzmacher> knows more about that. The
>    reason why these devices aren't supported is that they use a different
>    protocol.
>
> By the way, a link to Handy Techs NVDA page is already provided in the
> user guide.
>
> An additional issue I see is that the driver has a port parameter and the
> previous driver did not. So, people installing this and downgrading to an
> NVDA release containing the COM based driver will get an error and need to
> change nvda.ini manually. I have no idea how we can prevent this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-333983893>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KEn8v1q0wxihNz6HsnA03oRNJrQRks5soqbQgaJpZM4PT5As>
> .
>

bramd committed:
Small changes based on review by @michaelDCurran

 * Model._display is now a getter that uses a weakref under the hood
 * Demoted some log messages to debugWarning
 * Removed old TODO comment and clarified the GoHubs TODO comment
 * Documented _allSubclasses function

bramd commented:
@michaelDCurran:
> 1. log.warning is used a bit. Should that instead be debugWarning? Or was this deliberate?

The messages that were logged as warnings are unexpected (unknown packet types, unknown display models). That being said, unknown packet types are just ignored and should not cause issues and the RuntimeError that is raised for an unknown model will probably be logged anyway, so I demoted them to debugWarning.

> 2. There are a few comments about things to do. Do these need to be done before this goes to Next?

No, I clarified the only TODO comment left and removed one that wasn't really relevant anymore. The remaining TODO is about autodetection of a USB-serial chip used in older models by GoHubs. This displays should just work fine if the USB serial port is selected, but will not be autodetected. Probably @FelixGruetzmacher can help with this? I don't have any GoHubs based serial converters.

michaelDCurran commented:
If we renamed the driver, then config for any blind deaf person would 
break on upgrade. I think breaking on upgrade is worse than breaking on 
downgrade.

What exactly is the user experience  when downgrading?

We do inform the user always that downgrading is unsupported.

Commit:
Merge remote-tracking branch 'origin/master' into handytech-native

Commit:
Addressed review comment for hwIo. Now, the HID file handle will also be closed properly when HID initialisation fails after opening the handle but before calling super(Hid,self).__init__

bramd commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [Oct 4, 2017, 2:28 AM GMT+2](https://github.com/nvaccess/nvda/pull/7590#issuecomment-334017449 "2017-10-04T00:28:30Z - Replied by Github Reply Comments"):
> What exactly is the user experience when downgrading?

From what I've seen, driver fails to load because it tries to call __init__ with the port argument, which is not implemented. So, for the end-user there will be no braille and NVDA will fall back to the no braille driver.

FelixGruetzmacher commented:
Agreed, upgrade should not break Braille even temporarily, even if the
remedy is as simple as selecting a Braille driver in preferences.
Just an idea: What if the upgrade contained a dummy called handyTech which,
upon initialization, simply sets the driver to handyTechNative?

Bram Duvigneau <notifications@github.com> schrieb am Mi., 4. Okt. 2017 um
10:06 Uhr:

> *@michaelDCurran* <https://github.com/michaelDCurran> commented on Oct 4,
> 2017, 2:28 AM GMT+2
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-334017449>:
>
> What exactly is the user experience when downgrading?
>
> From what I've seen, driver fails to load because it tries to call *init*
> with the port argument, which is not implemented. So, for the end-user
> there will be no braille and NVDA will fall back to the no braille driver.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-334080120>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KMmO6FQSArVZ6OrJh0i6ta3VlyBUks5sozyCgaJpZM4PT5As>
> .
>

michaelDCurran commented:
Two other alternatives:
A. In setDisplayByName: if instantiation of newDisplay raises TypeError, 
we can try again without kwargs.
B. In setDisplayByName: Only provide kwargs to newDisplay if 
NewDisplay's getPossiblePorts does not raise NotImplementedError.

However, I think I am still leaning towards not caring about downgrades. 
It does not break NVDA, and the user (perhaps with help) does have a 
pretty easy way of fixing it.

leonardder commented:
[**@bramd**](https://github.com/bramd) commented on [4 okt. 2017 10:06 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-334080120 "2017-10-04T08:06:21Z - Replied by Github Reply Comments"):
> [**@michaelDCurran**](https://github.com/michaelDCurran) commented on [Oct 4, 2017, 2:28 AM GMT+2](https://github.com/nvaccess/nvda/pull/7590#issuecomment-334017449 "2017-10-04T00:28:30Z - Replied by Github Reply Comments"):
> > What exactly is the user experience when downgrading?
> 
> From what I've seen, driver fails to load because it tries to call __init__ with the port argument, which is not implemented. So, for the end-user there will be no braille and NVDA will fall back to the no braille driver.

This can be solved by resetting config to factory defaults, or removing the handy tech section from the configuration. IMO, these are things that are not a big problem to deal with, especially since downgrading is unsupported. Might be a bit tricky in the transmission period where this code is in next but not yet merged into master.

For #1271, I'm planning to fix this by using extensionPoints.callWithSupportedKwargs to initialize the driver. Of course, this isn't a fix for older versions.

leonardder commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [4 okt. 2017 11:58 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-334107583 "2017-10-04T09:58:38Z - Replied by Github Reply Comments"):
> Just an idea: What if the upgrade contained a dummy called handyTech which,
> upon initialization, simply sets the driver to handyTechNative?

We can also use a config upgrade step for this if this is really necessary.

michaelDCurran commented:
Err, ignore   all of my alternatives in the last comment. We're talking 
about previous copies of NVDA :)


However, my point still stands that I am leaning towards not caring for 
downgrades anyway.

leonardder commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -156,6 +158,14 @@ def postInit(self):
+	def _get__display(self):
+		"The L{BrailleDisplayDriver} which initialized this Model instance"
+		# self._displayRef is a weakref, call it to get the object
+		if self._displayRef and callable(self._displayRef):
```
Why this if check? Note that calling a weak reference returns None when the object died, so you can just do 			return self._displayRef()

leonardder commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -128,9 +129,9 @@ class Model(AutoPropertyObject):
+	#: A weak reference to the driver instance
 	#: @type; BrailleDisplayDriver
-	_display = None
+	_displayRef = None
```
While some model attributes are retrieved from the model class rather than an instance of the class, this attribute is only used for instances. I think setting self._displayRef to the weakref in __init__ is enough and would personally just remove this definition.

leonardder commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -425,6 +435,14 @@ class Modular80(Modular):
+
+	@param cls: the base class to list subclasses of
+	@type cls: class
+	@rtype list[class]
```
Shouldn't this be @rtype: [class], at least with a colon?

leonardder commented on source/brailleDisplayDrivers/handyTech.py:
```diff
@@ -477,7 +495,6 @@ class BrailleDisplayDriver(braille.BrailleDisplayDriver, ScriptableObject):
 
 	@classmethod
 	def check(cls):
-		# TODO: Probably return False if there is no serial port at all?
```
WIll be dealt with in #1271

bramd committed:
Fixed epydoc syntax in documentation for _allSubclasses

Based on review comment of @leonardder

bramd committed:
Merge branch 'handytech-native' of github.com:bramd/nvda into handytech-native

Commit:
* Remove useless if check from Model._get__display to get the referenced object, as self._displayRef is now always callable
* Surround some doc strings with triple quotes to stick with NVDA's convensions, even though the particular doc strings only take a single ine

Commit:
Fixed another doc string and no longer declare Model._displayRef as class attribute, as it is instance specific.

leonardder commented:
@michaeldcurran: I belief this is ready for another round of review.

michaelDCurran commented:
Reviewed and looks good. Is this ready to be incubated?

leonardder commented:
[**@michaelDCurran**](https://github.com/michaelDCurran) commented on [9 okt. 2017 07:49 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-335070770 "2017-10-09T05:49:52Z - Replied by Github Reply Comments"):
> Reviewed and looks good. Is this ready to be incubated?

To me, sure. Keep in mind this requires nvaccess/nvda-misc-deps#6

michaelDCurran commented:
Actually,


If it is just removals from miscDeps, we can do that at a later date. In 
fact, eventually SScons can also be removed. We're trying to get as much 
out of miscDeps as possible as it just causes extra administration.


Therefore, might I ask that you change the miscDeps submodule reference 
in this branch back to what is in NVDA master for this PR. Then we'll do 
a miscDeps clean up at a later date, but still before this release. Just 
once we switch to Windows 10 sdk.


Thanks

Commit:
Merge remote-tracking branch 'origin/master' into handytech-native

Commit:
Revert miscDeps to 3707b8e

Commit:
* HID Serial converter: verify incoming packets to make sure that the buffer always contains only one packet
* Make sure to handle old style ack properly, even though it is quite unexpected that they will ever arrive

leonardder commented:
@FelixGruetzmacher: Could you please review the most recent changes, particularly the ones at the top of onReceive? Also, it would be appreciated if you could test drive this new code with an Easy Braille or other HID serial converter based device.

The current driver in next doesn't handle overlapping packets properly, resulting in ack packages getting lost or key releases not being handled correctly. Packets in the buffer are now checked correctly, ensuring that handled packets are always complete.

As soon as you give your consent, I'll ask Mick to do another review.

FelixGruetzmacher commented:
Will do that tomorrow so we won't lose valuable time. I'm very very eager
to see this thing in 2017.4. You are doing such an amazing job.

Leonard de Ruijter <notifications@github.com> schrieb am Do., 12. Okt. 2017
um 15:39 Uhr:

> @FelixGruetzmacher <https://github.com/felixgruetzmacher>: Could you
> please review the most recent changes, particularly the ones at the top of
> onReceive? Also, it would be appreciated if you could test drive this new
> code with an Easy Braille or other HID serial converter based device.
>
> The current driver in next doesn't handle overlapping packets properly,
> resulting in ack packages getting lost or key releases not being handled
> correctly. Packets in the buffer are now checked correctly, ensuring that
> handled packets are always complete.
>
> As soon as you give your consent, I'll ask Mick to do another review.
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-336140167>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KGdiVWbcTm3QgXrh_nH6SCn5bwIDks5srhapgaJpZM4PT5As>
> .
>

FelixGruetzmacher commented:
Completed code inspection and all looks good except one potential issue: When a device id packet, the one consisting of 0xfe followed by the model id, is received from a hid to serial converter, the code currently does not check if there are more bytes beyond those two, so in theory, if that packet got lumped together with another one, that last one might get lost. Is this a ghost I'm seeing or actually relevant?

leonardder commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [16 okt. 2017 15:41 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-336889858 "2017-10-16T13:41:00Z - Replied by Github Reply Comments"):
> Completed code inspection and all looks good except one potential issue: When a device id packet, the one consisting of 0xfe followed by the model id, is received from a hid to serial converter, the code currently does not check if there are more bytes beyond those two, so in theory, if that packet got lumped together with another one, that last one might get lost. 

Hmm, very true indeed.

> Is this a ghost I'm seeing or actually relevant?

I think this should be considered. Is it true that non-extended packages are always two bytes in size?

Commit:
Properly handle multiple packets in the buffer when the first packet is non-extended

leonardder commented:
@MichaelDCurran: Could you please review the last few commits?

FelixGruetzmacher commented:
Yes, I believe this to be the case in all scenarios we are considering here. Also, just looked at your change and I believe this addresses it. Thank you.

leonardder commented:
@bramd and @FelixGruetzmacher, I wonder whether we should put a timer or a safety guard on the _awaitingAck property. To give an example:

1. Connect a Modular Evolution 88
2. Disable the device with the right side switch
3. Change focus, braille is sent to the device
4. awaitingAck is stuck in the True position. Setting it back to False raises a serial error and NVDA falls back to no braille

Note that this is currently not very relevant, as it is an edge case and the connection is lost when disabling and enabling the device anyway. However, when when #1271 gets into view, this is certainly an issue, as braille auto detection won't be triggered and thus the connection won't be reinstated.

We really should come up with a generic framework for protocols where we rely on ACKs

FelixGruetzmacher commented:
Maybe we should stop waiting for an ack after, say, 0.2 s, and increase a
missing_ack counter. If an ack is received at some point, that counter
should drop back to zero. If it reaches 3, meaning 3 consecutive missing
acks, that amounts to a connection loss. That stops the edge case from
causing trouble, but still honors the reason why an ack was put in place at
all, namely, detecting fatal errors.

Leonard de Ruijter <notifications@github.com> schrieb am Di., 17. Okt. 2017
um 09:05 Uhr:

> @bramd <https://github.com/bramd> and @FelixGruetzmacher
> <https://github.com/felixgruetzmacher>, I wonder whether we should put a
> timer or a safety guard on the _awaitingAck property. To give an example:
>
>    1. Connect a Modular Evolution 88
>    2. Disable the device with the right side switch
>    3. Change focus, braille is sent to the device
>    4. awaitingAck is stuck in the True position. Setting it back to False
>    raises a serial error and NVDA falls back to no braille
>
> Note that this is currently not very relevant, as it is an edge case and
> the connection is lost when disabling and enabling the device anyway.
> However, when when #1271 <https://github.com/nvaccess/nvda/issues/1271>
> gets into view, this is certainly an issue, as braille auto detection won't
> be triggered and thus the connection won't be reinstated.
>
> We really should come up with a generic framework for protocols where we
> rely on ACKs
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-337137999>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KMmZgaVcnCAifvTpapERc0gpB9Boks5stFHPgaJpZM4PT5As>
> .
>

leonardder commented:
[**@FelixGruetzmacher**](https://github.com/FelixGruetzmacher) commented on [17 okt. 2017 09:14 CEST](https://github.com/nvaccess/nvda/pull/7590#issuecomment-337139787 "2017-10-17T07:14:45Z - Replied by Github Reply Comments"):
> Maybe we should stop waiting for an ack after, say, 0.2 s, and increase a
> missing_ack counter. If an ack is received at some point, that counter
> should drop back to zero. If it reaches 3, meaning 3 consecutive missing
> acks, that amounts to a connection loss. That stops the edge case from
> causing trouble, but still honors the reason why an ack was put in place at
> all, namely, detecting fatal errors.

My problem with a timer is that it might, or might not, conflict with thread safety. I'd rather only work with a counter or with waitForRead. However, calling wiatForRead from the background thread simply has no effect, as it doesn't allow reading to take place while waiting. :P This is an issue that will be solved in #1271 as well. 

If you guys agree, I don't think this issue should block this driver from landing into a release. I've been working with this driver for weeks now and I haven't seen any missed ACKs except for the edge case I mentioned.

FelixGruetzmacher commented:
I stand convinced.

Leonard de Ruijter <notifications@github.com> schrieb am Di., 17. Okt. 2017
um 09:27 Uhr:

> *@FelixGruetzmacher* <https://github.com/FelixGruetzmacher> commented on 17
> okt. 2017 09:14 CEST
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-337139787>:
>
> Maybe we should stop waiting for an ack after, say, 0.2 s, and increase a
> missing_ack counter. If an ack is received at some point, that counter
> should drop back to zero. If it reaches 3, meaning 3 consecutive missing
> acks, that amounts to a connection loss. That stops the edge case from
> causing trouble, but still honors the reason why an ack was put in place at
> all, namely, detecting fatal errors.
>
> My problem with a timer is that it might, or might not, conflict with
> thread safety. I'd rather only work with a counter or with waitForRead.
> However, calling wiatForRead from the background thread simply has no
> effect, as it doesn't allow reading to take place while waiting. :P This is
> an issue that will be solved in #1271
> <https://github.com/nvaccess/nvda/issues/1271> as well.
>
> If you guys agree, I don't think this issue should block this driver from
> landing into a release. I've been working with this driver for weeks now
> and I haven't seen any missed ACKs except for the edge case I mentioned.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-337142562>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ARU1KOlFLv8Fd69jhjhi-scrlsExTBLTks5stFb0gaJpZM4PT5As>
> .
>

Commit:
Merge remote-tracking branch 'origin/master' into handytech-native

Commit:
Small typo fixes in user guide, added joystick keys from actilino

leonardder commented:
@michaeldcurran: I did some small typo fixes to the user guide yesterday and added some gestures that weren't yet in there.

I think the user guide is clear about the old devices (Modular with old firmware and Braillino) not being supported). But if you consider this to be to early for 2017.4, I can understand this. However, note that we should also consider that landing too much braille changes in one release increases the chance of unresolved bugs.

leonardder commented:
Actually, I think a good intermediate solution would be merging this pr while keeping the com server for now. This way, people only have to install the add-on provided by handy tech, and this is clearly documented.

zstanecic commented:
and... will this be merged into 2017.4?

it will be great, because this driver is more responsive than the com 
server in the MS word with braille.


W dniu 01.11.2017 o 08:29, Leonard de Ruijter pisze:
>
> Actually, I think a good intermediate solution would be merging this 
> pr while keeping the com server for now. This way, people only have to 
> install the add-on provided by handy tech, and this is clearly documented.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-341018144>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AKohk3k0nJD0QzabXf_vwTFqkT1xwoOoks5syB3KgaJpZM4PT5As>.
>

michaelDCurran commented:
In this case, we'd have to remove the changes to setup.py and readme.md 
from the pr. Is that correct?

If so, could someone do this please?

zstanecic commented:
so, michael, will this be merged in this release?


W dniu 01.11.2017 o 22:52, Michael Curran pisze:
> In this case, we'd have to remove the changes to setup.py and readme.md
> from the pr. Is that correct?
>
> If so, could someone do this please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nvaccess/nvda/pull/7590#issuecomment-341254270>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AKohkx-Ennx054_bCiO2KhA7oYeDPMc_ks5syOgGgaJpZM4PT5As>.
>

michaelDCurran commented:
Just a heads up that I will wait until tomorrow Friday 0 uTC to do 
translation freeze.
If this pr can be changed so that setup.py and readme.md reflect your 
proposal of keeping the COM server, I'll merge it to master.
As this PR will affect translations (changes.t2t at least), it needs to 
be merged before the freeze.

Commit:
Revert handy tech com server removal

leonardder commented:
There is a commit on the way where I reverted the removal of the com server. Note that I still removed one line from the readme, stating that you need the handy tech universal driver when running from source. This is no longer the case. You still need the handy tech universal driver installed when running from source with the Handy Tech add-on, but that's no longer our responsibility.

Commit:
Merge remote-tracking branch 'origin/master' into handytech-native

leonardder commented:
Ah, and there's another one, comInterfaces_sconscript. The handy tech com server python file is no longer delivered with NVDA, but built on demand again.
